src/controller/CHIPDeviceController.cpp

```diff
@@ -349,6 +350,45 @@ CHIP_ERROR DeviceController::GetPeerAddressAndPort(PeerId peerId, Inet::IPAddres
     return CHIP_NO_ERROR;
 }
 
+void DeviceController::OnPIDReadResponse(void * context, uint16_t value)
+{
+    ChipLogProgress(Controller, "Received PID for the device. Value %d", value);
+    DeviceController * controller       = reinterpret_cast<DeviceController *>(context);
+    controller->mSetupPayload.productID = value;
+
+    if (controller->OpenCommissioningWindowInternal() != CHIP_NO_ERROR)
+    {
+        OnOpenPairingWindowFailureResponse(context, 0);
+    }
+}
+
+void DeviceController::OnVIDReadResponse(void * context, uint16_t value)
+{
+    ChipLogProgress(Controller, "Received VID for the device. Value %d", value);
+
+    DeviceController * controller = reinterpret_cast<DeviceController *>(context);
+
+    controller->mSetupPayload.vendorID = value;
+
+    OperationalDeviceProxy * device =
+        controller->mCASESessionManager->FindExistingSession(controller->mDeviceWithCommissioningWindowOpen);
+    VerifyOrReturn(device != nullptr, ChipLogError(Controller, "Could not find device for opening commissioning window"));
+
+    constexpr EndpointId kBasicClusterEndpoint = 0;
+    chip::Controller::BasicCluster cluster;
+    cluster.Associate(device, kBasicClusterEndpoint);
+
+    Callback::Cancelable * successCallback = controller->mPIDReadCallback.Cancel();
+    Callback::Cancelable * failureCallback = controller->mVIDPIDReadFailureCallback.Cancel();
+    cluster.ReadAttributeProductID(successCallback, failureCallback);
```

Comment:
So we're generally trying to move away from the Callback-based APIs on clusters.  Is the fact that we're using Callback/Cancelable here something we critically rely on?  Or could we call the templated `ReadAttribute` API that takes just function callbacks instead?

Reply:
No, use of Callback is not necessary. Where can I find the templated `ReadAttribute` API?

Reply:
It's in `src/controller/CHIPCluster.h`.  Used like so:
```
   CHIP_ERROR err = ReadAttribute<Clusters::ClusterName::Attributes::AttrName::TypeInfo>(context, successCallback, failureCallback);
```